### PR TITLE
Always compute frozen partial documents in the BG to ensure any incurred IO is never on the UI thread.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Curious how we feel about this approach @jasonmalinowski

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 1]
```diff
@@ -1633,12 +1633,16 @@ private NonReentrantLock StateLock
...
         /// This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
         /// </summary>
-        public SolutionState WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, CancellationToken cancellationToken)
+        public async ValueTask<SolutionState> WithFrozenPartialCompilationIncludingSpecificDocumentAsync(DocumentId documentId, CancellationToken cancellationToken)
         {
             try
             {
+                // Ensure we explicitly move to the BG so that if we do any IO it does not potentially happen
+                // on a UI thread caller.
+                await Task.Yield().ConfigureAwait(false);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1642]: This is not necessary, and causes harm to the fast path. `GetSyntaxTreeAsync` is already asynchronous on the I/O paths.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 2]
```diff
@@ -1633,12 +1633,16 @@ private NonReentrantLock StateLock
...
+
                 var doc = this.GetRequiredDocumentState(documentId);
-                var tree = doc.GetSyntaxTree(cancellationToken);
+                var tree = await doc.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
 
                 using (this.StateLock.DisposableWait(cancellationToken))
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1647]: 📝 This should use the async form in asynchronous code

File: src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocument.cs [hunk 3]
```diff
@@ -24,13 +25,13 @@ internal SourceGeneratedDocument(Project project, SourceGeneratedDocumentState s
...
 
-        internal override Document WithFrozenPartialSemantics(CancellationToken cancellationToken)
+        internal override ValueTask<Document> WithFrozenPartialSemanticsAsync(CancellationToken cancellationToken)
         {
             // For us to implement frozen partial semantics here with a source generated document,
             // we'd need to potentially deal with the combination where that happens on a snapshot that was already
             // forked; rather than trying to deal with that combo we'll just fall back to not doing anything special
             // which is allowed.
-            return this;
+            return new(this);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocument.cs:34]: 💡 Use `ValueTaskFactory` for clarity

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1647]: This is a NonReentrantLock, which doesn't currently have the async codepath :)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1642]: good point.  I'm unsure what the best thing is to do here.  The issue is lower down (in the lock) with FreezePartialStateWithTree.  that's the part that can cause IO, and i'm trying to ensure we're not on the UI thread there.  thoughts on a better approach?

File: src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocument.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocument.cs]: done.

**Comment by CyrusNajmabadi** [comment]
Closing out for now.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs:1642]: @CyrusNajmabadi: so are we able to defer the actual creation of the compilation and fetching of symbols a bit further to the actual GetCompilationAsync() call? I think that's the right fix here.

**Pull request closed without merging**